### ISSUE TITLE
JDK 7 URL.openStream

### ISSUE THREAD

**Issue by gaellalire** [bcgit/bc-java#709]
I added `bcprov-jdk15to18-1.65.jar` and `bctls-jdk15to18-1.65.jar` in `$JAVA_HOME/jre/lib/ext`

Then I edited `$JAVA_HOME/jre/lib/security/java.security` to add
```
security.provider.1=org.bouncycastle.jce.provider.BouncyCastleProvider
security.provider.2=org.bouncycastle.jsse.provider.BouncyCastleJsseProvider
```
And a simple call to
```java
new URL("https://mydns.fr").openStream();
```

leads to this exception

```java
Exception in thread "main" org.bouncycastle.tls.TlsFatalAlert: certificate_unknown(46)
	at org.bouncycastle.jsse.provider.ProvSSLSocketDirect.checkServerTrusted(Unknown Source)
	at org.bouncycastle.jsse.provider.ProvTlsClient$1.notifyServerCertificate(Unknown Source)
	at org.bouncycastle.tls.TlsUtils.processServerCertificate(Unknown Source)
	at org.bouncycastle.tls.TlsClientProtocol.handleServerCertificate(Unknown Source)
	at org.bouncycastle.tls.TlsClientProtocol.handleHandshakeMessage(Unknown Source)
	at org.bouncycastle.tls.TlsProtocol.processHandshakeQueue(Unknown Source)
	at org.bouncycastle.tls.TlsProtocol.processRecord(Unknown Source)
	at org.bouncycastle.tls.RecordStream.readRecord(Unknown Source)
	at org.bouncycastle.tls.TlsProtocol.safeReadRecord(Unknown Source)
	at org.bouncycastle.tls.TlsProtocol.blockForHandshake(Unknown Source)
	at org.bouncycastle.tls.TlsClientProtocol.connect(Unknown Source)
	at org.bouncycastle.jsse.provider.ProvSSLSocketDirect.startHandshake(Unknown Source)
	at org.bouncycastle.jsse.provider.ProvSSLSocketDirect.startHandshake(Unknown Source)
	at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:559)
	at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:185)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1301)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:254)
	at java.net.URL.openStream(URL.java:1041)
	at Connect.main(Connect.java:12)
Caused by: java.security.cert.CertificateException: No subject alternative name found matching IP address 149.xx.xx.xx
	at org.bouncycastle.jsse.provider.HostnameUtil.checkHostname(Unknown Source)
	at org.bouncycastle.jsse.provider.ProvX509TrustManager.checkEndpointID(Unknown Source)
	at org.bouncycastle.jsse.provider.ProvX509TrustManager.checkEndpointID(Unknown Source)
	at org.bouncycastle.jsse.provider.ProvX509TrustManager.checkExtendedTrust(Unknown Source)
	at org.bouncycastle.jsse.provider.ProvX509TrustManager.checkTrusted(Unknown Source)
	at org.bouncycastle.jsse.provider.ProvX509TrustManager.checkServerTrusted(Unknown Source)
	... 19 more
```

I guess that is because `HostnameUtil.checkHostname` received a `149.xx.xx.xx` as hostname instead of `mydns.fr`. The verification by IP is not possible with Let's encrypt certificates, hence the exception.
I don't know if the code replacing DNS by IP is in BC or Sun code.

I think such a simple code should work.

I can think of :
- BC code could be changed to give the DNS name to HostnameUtil
- HostnameUtil don't check at all the hostname unless a system variable is set.
- your better fix here ...

**Comment by peterdettman** [comment]
Actually BCJSSE never sees the hostname and cannot access it reliably, which is the fault of the JDK code itself. This is basically a duplicate of https://github.com/bcgit/bc-java/issues/460 . The important part is:

> when using HttpsURLConnection SunJSSE uses some magic (reflection
and/or internal API) to tell the socket about the 'original hostname'
used for the connection ("google.com" here), and we can't use that same
magic

Specifically, HttpsURLConnection (via its internal HttpsClient) creates the socket using the default constructor (or at least without the hostname), then later calls setHost, but only for instances of sun.security.ssl.SSLSocketImpl (tested by reflection). Therefore our sockets never see the original hostname.

If stuck with HttpsURLConnection, then the second workaround I give in https://github.com/bcgit/bc-java/issues/460#issuecomment-461085449 is still the recommended solution, however it may be preferable to switch to an HTTPS implementation without these shortcomings (which to my knowledge is any other one).

**Comment by gaellalire** [comment]
OK, thanks for the informations and sorry for the duplicate.
I do not need a workaround, the software I ran is using Apache HttpClient. The URL.openStream was just a unit test.

However for someone who wants to execute a software which is calling URL.openStream and cannot run/port it to recent JVM there is a way.
The way is to do as I did (use BC), and create a class which call setURLStreamHandlerFactory to use Apache HttpClient for https before calling the real main.

Calling setURLStreamHandlerFactory should not be done by BC, so you don't have to fix anything.
